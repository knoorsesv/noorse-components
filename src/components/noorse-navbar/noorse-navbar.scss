@import '../../global/styles';

@mixin stick-header-to-top() {

  position: fixed;
  top: 0;
  left: 0;

}

@mixin fix-icon {
  .navbar-icon {
    z-index: 3;
    position: fixed;
    top: 20px;
    left: 20px;
    height: 20px;
    width: 20px;
    background-image: url('assets/menu-icon.svg');
  }
}

@mixin hide-icon() {
  .navbar-icon {
    display: none;

  }
}

@mixin float-menu-left() {

  .navbar-main {
    &:not(.menu-shown) {
      display: none;
    }

    &.menu-shown {
      display: inline-block;
      position: fixed;
      overflow: scroll;
      max-height: 50vh;
      flex-direction: column;
      top: 20px;
      left: 20px;
      padding-top: 20px;
      background-color: $neurse-groen;
    }
  }
}

.navbar {
  width: 100%;

  display: flex;
  justify-content: center;

  .navbar-main {
    display: flex;
    justify-content: center;
  }

  // big-screen
  @media screen and (min-width: $tablet) {
    @include hide-icon();

    &.in-sight {

      .navbar-dropdown {

      }
    }

    &.out-of-sight {
      @include stick-header-to-top();
    }

  }

  // small-screen
  @media screen and (max-width: $tablet) {

    &.in-sight {
      @include hide-icon();
    }

    &.out-of-sight {
      @include fix-icon();
      @include float-menu-left();
      min-height: 0;

    }

  }

  .navbar-dropdown {
    border-radius: $navbar-dropdown-radius;
  }

  .is-hoverable:hover {
    .navbar-dropdown {
      animation: fadeIn 1s;
      animation-timing-function: ease-out;
    }
  }

  .navbar-item {
    font-size: .8rem;
    @include white-text();
    a.is-active {
      &, &:hover {
        background-color: $neurse-groen-darker;
        color: $neurse-geel;
      }
    }
  }
}

@keyframes fadeIn {
  0% {
    opacity: 0;
  }

  100% {
    opacity: 1;
  }

}

