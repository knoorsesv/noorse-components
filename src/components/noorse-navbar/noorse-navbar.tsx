import {Component, Element, Event, EventEmitter, h, Listen, Prop, State} from "@stencil/core";
import {Navigation} from "../../types.interface";


@Component({
  tag: 'noorse-navbar',
  styleUrl: 'noorse-navbar.scss',
  shadow: true
})
export class NoorseNavbar {


  @Prop() navigation: string;
  @Event() itemSelected: EventEmitter;
  @State() navigationParsed: Navigation;
  @State() selectedItem: string;
  @Element() el: HTMLElement;
  @State() positionClass: string = "is-active";
  @State() showMenu: boolean = false;

  componentWillLoad() {
    if (this.navigation) this.navigationParsed = JSON.parse(this.navigation);
    this.setClassBasedOnScrollHeight();
  }

  @Listen('scroll', {target: 'document'})
  handleScroll() {
    this.setClassBasedOnScrollHeight();
  }


  private setClassBasedOnScrollHeight() {
    if (this.el.getBoundingClientRect().top < 0) {
      this.positionClass = "out-of-sight"
    } else if (this.el.getBoundingClientRect().top > 5) {
      this.positionClass = "in-sight"
    }
  }

  navigationItems() {
    return this.navigationParsed ?
      this.navigationParsed.items.map((item) => {
        return item.subItems ? this.getDropDownTag(item) : this.getLinkTag(item)
      }) : ''
  }

  private getDropDownTag(item) {
    return <div class="navbar-item has-dropdown is-hoverable">
      <a class="navbar-link navbar-item is-arrowless"> {item.name} </a>
      {this.dropDownItem(item.subItems)}
    </div>;
  }

  dropDownItem(subItems) {
    return <div class="navbar-dropdown">
      {subItems.map((subItem) => this.getLinkTag(subItem))}
    </div>
  }

  getLinkTag(item) {
    const itemClass = `navbar-item ${item.name === this.selectedItem ? "is-active" : ""}`;
    return <a class={itemClass}
              onClick={() => this.select(item)}>{item.name}</a>
  }

  select(item) {
    this.selectedItem = item.name;
    this.itemSelected.emit(item.name);
  }

  render() {
    return <div>
      <nav class={`navbar ${this.positionClass}`}>
        <div class={`navbar-icon`} onClick={() => this.showMenu = !this.showMenu}>        </div>
        <div class={`navbar-main ${this.showMenu ? "menu-shown": ""}`}>
          {
            this.navigationItems()
          }
        </div>
      </nav>
      {this.positionClass === 'out-of-sight' && <div class={`placeholder ${this.positionClass}`}></div>}
    </div>
  }
}



