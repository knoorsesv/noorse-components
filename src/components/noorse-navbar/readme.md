# noorse-header



<!-- Auto Generated Below -->


## Properties

| Property     | Attribute    | Description | Type     | Default     |
| ------------ | ------------ | ----------- | -------- | ----------- |
| `navigation` | `navigation` |             | `string` | `undefined` |


## Events

| Event          | Description | Type               |
| -------------- | ----------- | ------------------ |
| `itemSelected` |             | `CustomEvent<any>` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
