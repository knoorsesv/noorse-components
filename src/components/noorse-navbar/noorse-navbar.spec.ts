import {newSpecPage} from "@stencil/core/testing";
import {SpecPage} from "@stencil/core/dist/declarations";
import {NoorseNavbar} from "./noorse-navbar";

describe("Noorse Box", function () {

  let navbar: SpecPage;

  beforeEach(async () => {

    const navigation = JSON.stringify({
      items: [
        {name: 'whut'},
        {name: 'what', subItems: [{name: 'huehue'}]}
      ]
    });

    navbar = await newSpecPage({
      components: [NoorseNavbar],
      html: `<noorse-navbar navigation=${navigation}></noorse-navbar>`
    });
    await navbar.waitForChanges()
  });


  it('should render correct snapshots', function () {
    expect(navbar).not.toBeFalsy();
    expect(navbar.root).toMatchSnapshot()
  });

});
