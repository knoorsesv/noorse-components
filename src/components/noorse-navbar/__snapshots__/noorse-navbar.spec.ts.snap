// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Noorse Box should render correct snapshots 1`] = `
<noorse-navbar navigation="{&quot;items&quot;:[{&quot;name&quot;:&quot;whut&quot;},{&quot;name&quot;:&quot;what&quot;,&quot;subItems&quot;:[{&quot;name&quot;:&quot;huehue&quot;}]}]}">
  <mock:shadow-root>
    <div class="in-sight">
      <div class="navbar-icon">
        <img src="https://img.icons8.com/material/24/000000/menu.png">
      </div>
      <nav class="in-sight navbar">
        <div class="navbar-main">
          <a class="navbar-item">
            whut
          </a>
          <div class="has-dropdown is-hoverable navbar-item">
            <a class="is-arrowless navbar-item navbar-link">
              what
            </a>
            <div class="navbar-dropdown">
              <a class="navbar-item">
                huehue
              </a>
            </div>
          </div>
        </div>
      </nav>
    </div>
  </mock:shadow-root>
</noorse-navbar>
`;
